packages/backend-defaults/src/entrypoints/auth/helpers.ts

```diff
@@ -28,23 +28,31 @@ export function createCredentialsWithServicePrincipal(
   token?: string,
   accessRestrictions?: BackstagePrincipalAccessRestrictions,
 ): InternalBackstageCredentials<BackstageServicePrincipal> {
-  return Object.defineProperty(
-    {
-      $$type: '@backstage/BackstageCredentials',
-      version: 'v1',
-      principal: {
+  const result = {
+    $$type: '@backstage/BackstageCredentials',
+    version: 'v1',
+    principal: {
+      type: 'service',
+      subject: sub,
+      accessRestrictions,
+    },
+  } as const;
+  Object.defineProperty(result, 'token', {
+    enumerable: false,
+    configurable: true,
+    value: token,
+  });
+  Object.defineProperty(result, 'toString', {
+    enumerable: false,
+    configurable: true,
+    value: () =>
+      JSON.stringify({
+        $$type: '@backstage/BackstageCredentials',
         type: 'service',
         subject: sub,
```

Comment:
If we're gonna use it as a key where the same `toString` result means the same level of access - as it's intended to be used in TechDocs - I'd say we need to include `accessRestrictions` here too, either in plain form or a hash.